paddle/framework/recurrent_network_op.cc

```diff
@@ -0,0 +1,206 @@
+  size_t seq_len = dims[0];
+  auto& scopes = *step_scopes->GetMutable<std::vector<ScopePtr>>();
+  for (size_t step_id = 0; step_id < seq_len; step_id++) {
+    ScopePtr step_scope = scopes[step_id];
+    // TODO replace memorys' copy with reference
+    LinkMemories(scope, scopes, step_id);
+
+    net->GetMutable<PlainNet>()->Run(step_scope);
+  }
+
+  // prepare outputs
+  ConcateOutputs(scope);
+}
+
+void RecurrentOp::Init(const OpDesc& op_desc, AttributeMap& attrs) {
+  OperatorBase::Init(op_desc, attrs);
+  name_ = op_desc.name();
+  net_name_ = op_desc.name() + "_net";
+  step_scopes_name_ = op_desc.name() + "_step_scopes";
+  auto memories = GetAttr<std::vector<std::string>>("memories");
+  auto boot_memories = GetAttr<std::vector<std::string>>("boot_memories");
+  PADDLE_ENFORCE(memories.size() == boot_memories.size(),
+                 "The size of memories and boot_memories is mismatched.");
+  for (size_t i = 0; i < memories.size(); ++i) {
+    MemoryAttr mem_attr;
+    mem_attr.var = memories[i];
+    mem_attr.boot_var = boot_memories[i];
+    memory_attrs_.push_back(mem_attr);
+  }
+}
+
+void RecurrentOp::CreateScopes(ScopePtr scope) const {
+  auto dims = Input(scope, 0)->GetMutable<Tensor>()->dims();
+  size_t seq_len = dims[0];
+  Variable* scopes_var = scope->GetVariable(step_scopes_name_);
+  auto step_scopes = scopes_var->GetMutable<std::vector<ScopePtr>>();
+  // TODO Only two scopes are needed for inference, this case will be
+  // supported later.
+  if (seq_len > step_scopes->size()) {
+    for (size_t i = step_scopes->size(); i < seq_len; ++i) {
+      step_scopes->push_back(std::make_shared<Scope>(scope));
+    }
+  }
+}
+
+void RecurrentOp::CreateStepNet(ScopePtr scope) const {
+  Variable* var = scope->CreateVariable(net_name_);
+  auto step_net = GetAttr<std::string>("step_net");
+  // get the step net proto from the string.
+  // PADDLE_ENFORCE(
+  //   google::protobuf::TextFormat::ParseFromString(step_net,
+  //   &step_net_desc_));
+  // var->Reset<PlainNet>(new PlainNet(step_net_desc_));
+  // this is a fake net, it will be rewrite after the network has been merged.
+  NetDesc desc;
+  desc.name_ = "rnn_step_net";
+  var->Reset<PlainNet>(new PlainNet(desc));
+  // TODO add op descs
+}
+
+void RecurrentOp::SegmentInputs(ScopePtr scope) const {
+  Variable* scopes_var = scope->CreateVariable(step_scopes_name_);
+  auto& step_scopes = *scopes_var->GetMutable<std::vector<Scope*>>();
+
+  auto dims = Input(scope, 0)->GetMutable<Tensor>()->dims();
+  int seq_len = dims[0];
+  int batch_size = dims[1];
+  int dim = dims[2];
+  int length = batch_size * dim;
+  for (size_t i = 0; i < inputs_.size(); i++) {
+    const float* scope_input =
+        Input(scope, i)->GetMutable<Tensor>()->data<float>();
+    for (int j = 0; j < seq_len; j++) {
+      std::string name =
+          name_ + "@input_" + inputs_[i] + "@step_" + std::to_string(j);
+      Variable* input_var = step_scopes[j]->CreateVariable(name);
```

Comment:
The name is not correct. The name should be the same to each step. Maybe the `name_` is enough. This name is related to the operator's input name in the step net, which is specified in the protobuf description.